src/Internal/Service/PaymentProcessingService.php

```diff
@@ -64,23 +67,62 @@ public function process_payment( int $order_id, bool $manual_capture = false ) {
 		// Start with a basis context.
 		$context = $this->create_payment_context( $order_id, $manual_capture );
 
-		$request         = new PaymentRequest( $this->legacy_proxy );
-		$initial_state   = $this->state_factory->create_state( InitialState::class, $context );
-		$completed_state = $initial_state->process( $request );
+		$request       = new PaymentRequest( $this->legacy_proxy );
+		$initial_state = $this->state_factory->create_state( InitialState::class, $context );
 
-		return $completed_state;
+		return $initial_state->start_processing( $request );
 	}
 
 	/**
 	 * Instantiates a new empty payment context.
 	 *
-	 * @param int  $order_id       ID of the order that the context belongs to.
+	 * @param int  $order_id ID of the order that the context belongs to.
 	 * @param bool $manual_capture Whether manual capture is enabled.
+	 *
 	 * @return PaymentContext
 	 */
 	protected function create_payment_context( int $order_id, bool $manual_capture = false ): PaymentContext {
 		$context = new PaymentContext( $order_id );
 		$context->toggle_manual_capture( $manual_capture );
+
 		return $context;
 	}
+
+	/**
+	 * Get redirect URL when authentication is required (3DS).
+	 *
+	 * @param WC_Payments_API_Abstract_Intention $intent Intent object.
+	 * @param int                                $order_id Order id.
+	 *
+	 * @return string
+	 */
+	public function get_authentication_redirect_url( $intent, int $order_id ) {
+		$next_action = $intent->get_next_action();
+
+		if ( isset( $next_action['type'] ) && 'redirect_to_url' === $next_action['type'] && ! empty( $next_action['redirect_to_url']['url'] ) ) {
+			return $next_action['redirect_to_url']['url'];
+		}
+
+		$client_secret = $intent->get_client_secret();
+
+		if ( $this->legacy_proxy->call_static( \WC_Payments_Features::class, 'is_client_secret_encryption_enabled' ) ) {
+			$client_secret = $this->legacy_proxy->call_function(
+				'openssl_encrypt',
+				$client_secret,
+				'aes-128-cbc',
+				substr( $intent->get_customer_id(), 5 ),
+				0,
+				str_repeat( 'WC', 8 )
+			);
+		}
+
+		return sprintf(
+			'#wcpay-confirm-%s:%s:%s:%s',
+			substr( $intent->get_id(), 0, 2 ), // intents starts with pi_ or si_ so we need only to first two letters.
```

Comment:
Instead of string operations here, could we use the intent class?
```suggestion
			$intent instanceof WC_Payments_API_Setup_Intention ? 'si' : 'pi',
```